### PR TITLE
Track stats from the Deactivation Modal

### PR DESCRIPTION
**Pull request by leogermani**
Track stats from the Deactivation Modal actions

#### Changes proposed in this Pull Request:

* Adds tracking to actions in the Deactivation modal (open the dialog, cancel and confirm)

#### Is this a new feature or does it add/remove features to an existing part of Jetpack?
* This is part of the "Handle disconnections gracefully" project - p1HpG7-8VJ-p2

#### Does this pull request change what data or activity we track or use?

Yes, we are now tracking:
* When the user clicks "Disconnect and Deactivate", in order to deactivate the plugin (This only shows up if there's at least one other plugin relying on the Jetpack connection
* When the user confirms the deactivation
* When the user cancels the deactivation

#### Feedback wanted

* Are these the best event names? I'm using the same events used in the termination dialog in the Jetpack Dashboard. It already has these `location` and `purpose` that were meant to be used to differentiate when we reused this modal in the deactivation context (which we didn't do. We created another dialog from scratch)
* On the termination dialog in the Jetpack Dashboard there is no specific action to track the click to open the dialog. I kept the same standard.
* The deactivation click triggers a `window.location.href` change, it will take the user away from that page. Does it mean it could cancel the tracking that is fired just before?
* I'm not sure how to test it. I can confirm the function is being triggered correctly and there are no errors in the console, but I don't see the stats showing up on Tracks

#### Testing instructions:
* Set up a site using this branch and connect Jetpack
* Visit Plugins page on the Admin dashboard
* Check and see that the "Deactivate" link for the Jetpack plugin looks the same and nothing changed
* Create a file `wp-content/plugins/jetpack-connection-test-1.php` with following content:
```
<?php
/** Plugin Name: Jetpack Connection Test Plugin (1) */
add_action( 'plugins_loaded', function() {
	( new Automattic\Jetpack\Config() )->ensure( 'connection', array( 'slug' => 'jetpack-connection-test-1', 'name' => 'Jetpack Connection Test Plugin (1)' ) );
}, 1 );
```
* Activate this new plugin
* Now check that the label to deactivate Jetpack says "Disconnect and Deactivate"
* Click "Disconnect and Deactivate" and check that it opens a modal looking like the screenshots below
* Click Cancel and the close button at the top and confirm it closes the window
* Click the red "Disconnect and Deactivate" button and confirm it deactivates Jetpack

Now visit Tracks Live View and search for `jetpack_termination_dialog_%`

Make sure you see the sequence of events you've just performed (open, close, open terminate)

#### Proposed changelog entry for your changes:

* Tracks deactivation modal events

### CODE REVIEW

File: _inc/jetpack-deactivate-dialog.js [hunk 1]
```diff
@@ -33,19 +33,29 @@
...
+		purpose: 'deactivate',
+	};
+
 	deactivateLinkElem.attr( 'href', 'plugins.php#TB_inline?inlineId=jetpack_deactivation_dialog' );
 	deactivateLinkElem.attr( 'title', deactivate_dialog.title );
 	deactivateLinkElem.addClass( 'thickbox' );
 	deactivateLinkElem.html( deactivate_dialog.deactivate_label );
 	deactivateLinkElem.on( 'click', function( e ) {
 		observer.observe( body, { childList: true } );
+		analytics.tracks.recordEvent( 'jetpack_wpa_click', {
```
Comment (dereksmart) [inline _inc/jetpack-deactivate-dialog.js:47]: I've never been a fan of `jetpack_wpa_click` as a top-level event name used over all. It's hard to search for a prop only. Let's update the event name to `jetpack_click_deactivate_plugin` (if that's what it's actually tracking)

Comment (dereksmart) [reply _inc/jetpack-deactivate-dialog.js:47]: shoudl probably use the same prefix as the rest, so `jetpack_termination_dialog_something` for everything happening in this context

Comment (leogermani) [reply _inc/jetpack-deactivate-dialog.js]: changed to `jetpack_termination_dialog_open`

File: class.jetpack.php [hunk 2]
```diff
@@ -4083,6 +4083,14 @@ public function deactivate_dialog( $hook ) {
 					true
 				);
 
+				wp_enqueue_script(
```
Comment (dereksmart) [inline class.jetpack.php:4086]: Do we still need this?

**Comment by dereksmart** [comment]
General question: Why are we calling these events `termination` when everywhere else in the code we're calling it `deactivation`?

File: _inc/jetpack-deactivate-dialog.js [hunk 3]
```diff
@@ -33,19 +33,32 @@
...
 	deactivateLinkElem.attr( 'href', 'plugins.php#TB_inline?inlineId=jetpack_deactivation_dialog' );
 	deactivateLinkElem.attr( 'title', deactivate_dialog.title );
 	deactivateLinkElem.addClass( 'thickbox' );
 	deactivateLinkElem.html( deactivate_dialog.deactivate_label );
 	deactivateLinkElem.on( 'click', function( e ) {
 		observer.observe( body, { childList: true } );
+		window.jpTracksAJAX.record_ajax_event( 'termination_dialog_open', 'click', tracksProps );
 	} );
 
 	$( '#jetpack_deactivation_dialog_content__button-cancel' ).on( 'click', function( e ) {
```
Comment (dereksmart) [inline _inc/jetpack-deactivate-dialog.js:50]: By only watching the cancel button, we're missing out on the other events that close the modal. 
- X button.
- Clicking outside of the box. 

If we can catch all of these, I think it's reasonable to group them all under the same "dialog_close" event.

Comment (leogermani) [reply _inc/jetpack-deactivate-dialog.js:50]: At first I could not capture these events because thickbox is not very extensible. Will try again

Comment (leogermani) [reply _inc/jetpack-deactivate-dialog.js:74]: done! Bonus: also track closing using ESC key

**Comment by leogermani** [comment]
> General question: Why are we calling these events `termination` when everywhere else in the code we're calling it `deactivation`?

The event name was created in the Disconnect dialog, in our Dashboard, I kept the same name so the events are grouped (`location` and `purpose` props differentiate them)

File: class.jetpack.php (hunk 2 above)
Comment (leogermani) [reply class.jetpack.php]: no!

**Comment by dereksmart** [comment]
> I kept the same name so the events are grouped

Ok, makes enough sense. What threw me off is that this modal is right now specifically designed only for warning users about disrupting other plugins connections, and is not firing on every deactivation attempt. 

We will likely want to rethink this later when we revisit the deactivation modal and show it on every attempt. Going forward, it would be useful to be able to answer 
- How many times is someone deactivating/disconnecting one connected plugin when others are connected?
- How successful is this message in preventing accidental breakage of other plugin connections?

**Comment by leogermani** [comment]
Derek reported that there was a long delay after confirming the deactivation.

That's because it waits for the ajax call that tracks the event to complete before actually deactivating the plugin.

I thought it would be a good idea to disable the button while the ajax call is being made, to give user a bit of feedback and to avoid multiple clicks.

I've done that and need an ok for this again.

**Comment by jeherve** [comment]
> That's because it waits for the ajax call that tracks the event to complete before actually deactivating the plugin.

Yes, I've experienced the same thing when clicking on items in the Masterbar. I think it makes for a not-great user experience.

> I thought it would be a good idea to disable the button while the ajax call is being made, to give user a bit of feedback and to avoid multiple clicks.

That's a good way to mitigate that, nice call 👍

File: class.jetpack.php [hunk 4]
```diff
@@ -4072,13 +4072,30 @@ public function deactivate_dialog( $hook ) {
...
+					true
+				);
+
+				wp_register_script(
+					'jp-tracks-functions',
+					plugins_url( '_inc/lib/tracks/tracks-callables.js', JETPACK__PLUGIN_FILE ),
+					array( 'jp-tracks' ),
+					JETPACK__VERSION,
+					false
+				);
```
Comment (jeherve) [inline class.jetpack.php:4090]: Could we take that opportunity to register those scripts only once here:
https://github.com/Automattic/jetpack/blob/d25f37a67b0074df3be96536c963601dae89438b/class.jetpack.php#L1157-L1164

And then only add `jp-tracks-functions` as dependency here like you've done, and here:
https://github.com/Automattic/jetpack/blob/3010b47431d784ba4771188ca1da2637bb922d97/class.jetpack-idc.php#L266-L320

Comment (leogermani) [reply class.jetpack.php:4090]: Yes!

I was planning exactly that but thought I should do it in another PR and make the other places where these scripts are enqueued use this approach

**Comment by jeherve** [comment]
Cherry-picked to `branch-8.6` in 6a6ac58d1ea0fce489d6a3c9915311590a66d38c

**Pull request merged**